### PR TITLE
8324972: (bf) Make DirectByteBuffer.Deallocator idempotent

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to make deallocators and unmappers for memory regions idempotent. This is to prevent (likely very rare) duplicate invocations. 

There are no unit tests but it should be noted that the idempotent behavior (now correct) is similar to the  intended behavior before cf74b8c2a32f33019a13ce80b6667da502cc6722 but where idempotency was not guaranteed in a multi-threaded environment.

Passes tier1, 2, and 3 tests.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324972](https://bugs.openjdk.org/browse/JDK-8324972): (bf) Make DirectByteBuffer.Deallocator idempotent (**Bug** - P4) ⚠️ Issue is not open.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17647/head:pull/17647` \
`$ git checkout pull/17647`

Update a local copy of the PR: \
`$ git checkout pull/17647` \
`$ git pull https://git.openjdk.org/jdk.git pull/17647/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17647`

View PR using the GUI difftool: \
`$ git pr show -t 17647`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17647.diff">https://git.openjdk.org/jdk/pull/17647.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17647#issuecomment-1918643127)

### CODE REVIEW

**Comment by AlanBateman** [comment]
I'm busy this week but I will reply to this PR next week. I don't object to making it idempotent but I have concerns with using AtomicBoolean and it pulling in MH/VH classes. It very important that we have as few dependences as possible in this area.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 1]
```diff
@@ -75,15 +76,32 @@ class Direct$Type$Buffer$RW$$BO$
 
 #if[byte]
 
-    private record Deallocator(long address, long size, int capacity) implements Runnable {
-        private Deallocator {
+    private static final class Deallocator implements Runnable {
+
+        private static final AtomicIntegerFieldUpdater<Deallocator> UPDATER =
```
Comment (minborg) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:81]: It would be possible to use `Unsafe` directly instead of AIFU.

Comment (shipilev) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:81]: +1, just use `Unsafe` directly here. `AIFU` would probably even worse than `AtomicInteger` if we want to minimize dependencies. We already use Unsafe in adjacent classes here.

File: src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java [hunk 2]
```diff
@@ -47,6 +47,7 @@
 import java.nio.channels.SelectableChannel;
 import java.nio.channels.WritableByteChannel;
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
```
Comment (shipilev) [inline src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java:50]: Unused import?

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 3]
```diff
@@ -75,15 +77,31 @@ class Direct$Type$Buffer$RW$$BO$
 
 #if[byte]
 
-    private record Deallocator(long address, long size, int capacity) implements Runnable {
-        private Deallocator {
+    private static final class Deallocator implements Runnable {
+
+        private static final long INVOKED_OFFSET = Unsafe.getUnsafe()
```
Comment (msheppar) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:82]: There are two invocations of Unsafe.getUnsafe  introduced here, while the base class Buffer has initialised the static UNSAFE (i.e. the reference used for freeMemory.)   Could this static be used to replace Unsafe.getUnsafe ?

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template [hunk 4]
```diff
@@ -75,15 +77,31 @@ class Direct$Type$Buffer$RW$$BO$
...
+            this.address = address;
+            this.size = size;
+            this.capacity = capacity;
         }
 
         public void run() {
-            UNSAFE.freeMemory(address);
-            Bits.unreserveMemory(size, capacity);
+            // Ensure idempotency (paranoia)
+            if (Unsafe.getUnsafe().compareAndSetInt(this, INVOKED_OFFSET, 0, 1)) {
```
Comment (msheppar) [inline src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:99]: a question for my edification: is this compound statement suitably atomic to ensure synchronzied serial access to freeMemory block ?  There's no danger of concurrent interleaved execution resulting in inconsistent view of the member variable via the reflective  access ?

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template (hunk 3 above)
Comment (minborg) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:82]: Good catch!

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template (hunk 4 above)
Comment (minborg) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template:99]: The CAS operation establishes a happens-before relation between the CAS operation itself and any following operations (such as `freeMemory()`). Also, only one thread will succeed in changing the value from zero to one.

Comment (minborg) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: The behavior in this PR is now like the intended behavior in Java 21 (and before) but thread-safe. So, by induction, this should be as safe or safer as we had in <= 21. Adding blocking might introduce other problems in a multi-threaded environment.

**Comment by AlanBateman** [comment]
There is no explicit free for direct buffers (or unmap for mapped buffers) and the Cleaner action to free/unmap will only be run once in response to reference processing. The only scenario that I can come up is using Unsafe.invokeCleaner where it looks like we might be missing a RF to keep the DBB alive. I will look further to see if I can see anything else.

File: src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template (hunk 4 above)
Comment (msheppar) [reply src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template]: thanks for the replies, so all invocations lead to Atomic::cmpxchg, where some magic happens....I didn't realise that AtomicInteger is basically a wrapper around what has been added to Deallocator

**Comment by AlanBateman** [comment]
Just to add that I don't think Unsafe.invokeCleaner racing with the Cleaner is an issue. The clean method is specified to specified to invoke the cleaning action at most once, and the PhamtomCleanable.clean method appears to be correct.

I wonder if @msheppar could re-phrase JDK-8324972 to more clearly explain why there is perceived issue issue. Note that I'm not opposed to making the cleaner action idempotent, I'm just trying to see how it possible for it to run more than once.

**Comment by msheppar** [comment]
https://bugs.openjdk.org/browse/JDK-8316337  made a change advocating thread safety.  That change had a flaw, incorrectly viewing Unsafe.freeMemory as idempotent, and as such is not thread safe, nor is it safe to multiple sequential invocation of the run method. 
The logic in the original incarnation of the run method, protected against multiple invocations (sequential or concurrent), and in particular multiple invocation of the lower level native  free (memory) function, encapsulated in the Unsafe.freeMemory abstraction.
The issue is raised to re-instate those semantics - idempotency, and additionally thread safety. 
The raised issue is about correct behaviour in the context of multiple concurrent and sequential access, regardless of whether there is the possibility of one thread or multiple threads.
Per's retitle of the issue is a good one,  appropriately elaborating the semantics of the change

**Comment by AlanBateman** [comment]
> https://bugs.openjdk.org/browse/JDK-8316337 made a change advocating thread safety. That change had a flaw, incorrectly viewing Unsafe.freeMemory as idempotent, and as such is not thread safe, nor is it safe to multiple sequential invocation of the run method. 

The comments added to JDK-8316337 aren't very clear.  Would it be possible to outline the scenario that leads to the cleaning action be run more than once? I'm trying to see if there is a bug in the cleaner that explains this.

**Comment by msheppar** [comment]
Is the Deallocator abstraction designed to be thread safe as advocated in https://bugs.openjdk.org/browse/JDK-8316337  ?
Is it  correct and proper design, hence good practice, as per the original logic, that Deallocator guard against improper release of its encapsulated native memory more than once?

**Comment by AlanBateman** [comment]
> Is the Deallocator abstraction designed to be thread safe as advocated in https://bugs.openjdk.org/browse/JDK-8316337 ? Is it correct and proper design, hence good practice, as per the original logic, that Deallocator guard against improper release of its encapsulated native memory more than once?

Deallocator is a function that is intended to be run at most once. Mostly it will be called from the Cleaner thread but there is backdoor with Unsafe.invokeCleaner so it may be run by some other thread. Cleaner is designed, specified, and implemented, to invoke the cleaning action at most once, so Deallocator.run should run at most once. If there is a scenario where Deallocator.run can be run more than once then it is serious bug. I'm not opposed to make it idempotent but I'm concerned this is masking an issue elsewhere. If you have identified a scenario where it can be run more than once then it would be useful to get an outline so it can be studied.

**Comment by AlanBateman** [comment]
@msheppar Do you have any other information to add? I'm wondering if this issue should be closed as it seems to be passed on the assumption that the cleaning action runs more than once.

**Comment by msheppar** [comment]
> @msheppar Do you have any other information to add? I'm wondering if this issue should be closed as it seems to be passed on the assumption that the cleaning action runs more than once.

I've nothing to add other than to re-iterate if https://bugs.openjdk.org/browse/JDK-8316337 is made based on thread safety, then it needs to take into account the idempotent release of the underlying native memory as per the original logic. Otherwise base on your assertions the https://bugs.openjdk.org/browse/JDK-8316337 was not required

**Comment by AlanBateman** [comment]
> I've nothing to add other than to re-iterate if https://bugs.openjdk.org/browse/JDK-8316337 is made based on thread safety, then it needs to take into account the idempotent release of the underlying native memory as per the original logic. Otherwise base on your assertions the https://bugs.openjdk.org/browse/JDK-8316337 was not required

JDK-8316337 was about visibility and safe publication, it didn't introduce a way for the cleaning action to run more than once. I think we can close 8324972 and this PR.

**Comment by dfuch** [comment]
Cleaner and Cleanable are very clear that the cleaning action will be invoked at most once.
https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/ref/Cleaner.html

As long as the Deallocator cannot be accessed from anywhere else (and that seems the case), it doesn't really need to be idempotent. Making it idempotent is a kind of defense in depth fix - that strictly speaking is not needed. If Cleaner/Cleanable could allow the cleaning action to be invoked twice that would be a major bug - as according to the specification, this should not happen, and therefore most of cleaning actions wouldn't guard against it.
If the Deallocator was accessible from anywhere else than the cleaning action then making it idempotent would indeed be necessary.

**Comment by msheppar** [comment]
forget concurrency, forget idempotency, forget safe publication across threads etc.
Ask the question what is the Deallocator state representing ... it is encapsulating a  native memory address that should be freed by an os::free, and when that has been successfully executed, its state should show that referenced address should no longer be freed, becuse that potential side effects.
That's a simple and precise propsition.

Why are you trying to justify the use of a record to represent Deallocator state as immutable, when that state is not immutable

**Comment by AlanBateman** [comment]
> forget concurrency, forget idempotency, forget safe publication across threads etc. Ask the question what is the Deallocator state representing ... it is encapsulating a native memory address that should be freed by an os::free, and when that has been successfully executed, its state should show that referenced address should no longer be freed, becuse that potential side effects. That's a simple and precise propsition.

FileChannelImpl.Closer, NativeBuffer.Deallocator, and the task returned by DatagramChannelImpl.releaserFor are other examples of cleaning actions in this area. They aren't mutable objects. Instead, they are just functions that run at most once to release the resource.

**Comment by minborg** [comment]
I think we have concluded that the Cleaner can only run items once and that there is no other way the closers can run. So, with this in mind, can you resolve the issue as "Not an issue" @msheppar ?

The old "if (address == 0)" must be old code from before the Cleaner era.

**Comment by msheppar** [comment]
which by inference means that change https://bugs.openjdk.org/browse/JDK-8316337 was not required, also.
No need for threads safety or safe publication if there is no concurrency involved

I can close it but I don't agree with the conclusion. The Deallocator should convey whether the native memory block has been freed or not, and maintained its idempotent characteristic, irrespective of concurrency or at most once semantics. The latter  being wholely dependent on correct execution of Cleaner mechanism. At most once semantics should come from the implementation and the current implementation does not exhibit at most once semantics, while the original logic did provide at most once semantics for serial execution. But not for concurrent execution

**Comment by AlanBateman** [comment]
> which by inference means that change https://bugs.openjdk.org/browse/JDK-8316337 was not required, also. No need for threads safety or safe publication if there is no concurrency involved

There is concurrency. The DBB is allocated in one thread, the memory backing that DBB is freed by the Cleaner thread.


> I can close it but I don't agree with the conclusion. The Deallocator should convey whether the native memory block has been freed or not, and maintained its idempotent characteristic, irrespective of concurrency or at most once semantics. The latter being wholely dependent on correct execution of Cleaner mechanism. At most once semantics should come from the implementation and the current implementation does not exhibit at most once semantics, while the original logic did provide at most once semantics for serial execution. But not for concurrent execution

Let's close it.  "At most once" is clearly specified and avoids having every cleaning action (Deallocator is just one of many in the JDK) to have to defend against being called more than once.  By "original logic" then I assume you mean the unsafe check for address == 0 and "Paranoia" comment. This pre-dates the migration to the new Cleaner and should probably have been removed a long time to avoid any confusion that it could be called more than once.

**Comment by minborg** [comment]
The https://bugs.openjdk.org/browse/JDK-8316337 was made to ensure visibility and safe publication between, on the one hand, the thread that allocated the resource and, on the other hand, the distinct cleaner thread.

**Comment by msheppar** [comment]
ok, well that's a bit of a subtle one.  So you are saying that passing a reference to a  Cleaner of the newly constructed Dealloactor doesn't guarantee that Cleaner has a consistent state of the Deallocator, as shown in the code below. That's a bit dubious, as that would imply that every Factory class, or factory method, or object Builder would suffer the same publication effect, when they hand out constructed objects ?

        try {
            cleaner = Cleaner.create(this, new Deallocator(base, size, cap));
        } catch (Throwable t) {
            // Prevent leak if the Deallocator or Cleaner fail for any reason
            UNSAFE.freeMemory(base);
            Bits.unreserveMemory(size, cap);
            throw t;
        }

**Comment by minborg** [comment]
I am not sure I understand your question. Previously, before we changed to a record. the `address` field was not `final` and hence, we had no safe publication.

**Comment by msheppar** [comment]
I'm refering to the safe publication of Deallocator via the Cleaner.create code ... the Deallocator reference is stored in a final field in the Cleaner, and the Cleaner is published via a synchronzied method, which I think would represent safe publication of the Deallocator (prior to it being turned into a record) ?

**Pull request closed without merging**